swarm/storage/ldbstore.go

```diff
@@ -805,12 +805,12 @@ func (s *LDBStore) Put(ctx context.Context, chunk Chunk) error {
 	gcIdxKey := getGCIdxKey(&index)
 	gcIdxData := getGCIdxValue(&index, po, chunk.Address())
 	s.batch.Put(gcIdxKey, gcIdxData)
-	s.lock.Unlock()
 
 	select {
 	case s.batchesC <- struct{}{}:
 	default:
 	}
+	s.lock.Unlock()
```

Comment:
why?

Reply:
It is strange, but this is the data race: https://github.com/ethersphere/go-ethereum/issues/1198#issuecomment-461775049.

Reply:
I am puzzled here. can you explain?

Reply:
Closing s.batchesC is reported as write in https://github.com/ethersphere/go-ethereum/issues/1198#issuecomment-461775049 data race, and the writing to it in a different goroutine (this code) is causing the race condition. It is a data race, which operation would be first. This is an example where the ownership of the channel is shared to two or more goroutines. In this case, the solution can be with already existing mutex, while keeping the channel ownership shared.

Reply:
Interesting. Does this mean the channel might be closed before we try to write to it?

Note: If yes, then the channel will panic even in a `select` block.
https://play.golang.org/p/xTlQ6LYkQ8L

Reply:
We did not gat any such panics, but it is possible in general.

Reply:
To be honest I'm a little worried about that possible panic. As the race detector exactly showed the possibility. Now we just hide that case with locks, but it can still happen.

If nobody else is worried about this I let go.

Reply:
Yes, I am a bit worried also, but I also worry not to introduce some new bugs in storage codebase with more changes since we did not get this panic so far.

Reply:
lets just not close the channel ok?